packages/history/history.ts

```diff
@@ -0,0 +1,983 @@
+
+  /**
+   * Pushes a new location onto the history stack, increasing its length
+   * by one. If there were any entries in the stack after the current one,
+   * they are lost.
+   *
+   * @param to - The new URL
+   * @param state - Data to associate with the new location
+   */
+  push(to: To, state?: State): void;
+
+  /**
+   * Replaces the current location in the history stack with a new one.
+   * The location that was replaced will no longer be available.
+   *
+   * @param to - The new URL
+   * @param state - Data to associate with the new location
+   */
+  replace(to: To, state?: State): void;
+
+  /**
+   * Navigates `n` entries backward/forward in the history stack relative
+   * to the current index. For example, a "back" navigation would use go(-1).
+   *
+   * @param n - The delta in the stack index
+   */
+  go(n: number): void;
+
+  /**
+   * Navigates to the previous entry in the stack. Identical to go(-1).
+   *
+   * Warning: if the current location is the first location in the stack,
+   * this will unload the current document.
+   */
+  back(): void;
+
+  /**
+   * Navigates to the next entry in the stack. Identical to go(1).
+   */
+  forward(): void;
+
+  /**
+   * Sets up a listener that will be called whenever the current location
+   * changes.
+   *
+   * @returns unlisten - A function that may be used to stop listening
+   */
+  listen(listener: Listener<S>): () => void;
+
+  /**
+   * Prevents the current location from changing and sets up a listener
+   * that will be called instead.
+   *
+   * @returns unblock - A function that may be used to stop blocking
+   */
+  block(blocker: Blocker<S>): () => void;
+}
+
+/**
+ * A browser history stores the current location in regular URLs in a
+ * web browser environment. This is the standard for most web apps and
+ * provides the cleanest URLs the browser's address bar.
+ */
+export interface BrowserHistory<S extends State = State> extends History<S> {}
+
+/**
+ * A hash history stores the current location in the fragment identifier
+ * portion of the URL in a web browser environment.
+ *
+ * This is ideal for apps that do not control the server for some reason
+ * (because the fragment identifier is never sent to the server), including
+ * some shared hosting environments that do not provide fine-grained controls
+ * over which pages are served at which URLs.
+ */
+export interface HashHistory<S extends State = State> extends History<S> {}
+
+/**
+ * A memory history stores locations in memory. This is useful in stateful
+ * environments where there is no web browser, such as node tests or React
+ * Native.
+ */
+export interface MemoryHistory<S extends State = State> extends History<S> {
+  index: number;
+}
+
+type HistoryState = {
+  usr?: State;
+  key?: string;
+  idx: number;
+};
+
+const BeforeUnloadEventType = 'beforeunload';
+const HashChangeEventType = 'hashchange';
+const PopStateEventType = 'popstate';
+
+const readOnly: (obj: any) => any = __DEV__
```

Comment:
Hey, not sure if you saw this, but I left a comment in the original [TS PR](https://github.com/ReactTraining/history/pull/774) about whether this type is correct: https://github.com/ReactTraining/history/pull/774#discussion_r376317031